Upgrade EUI to v91.3.1

cee-chen opened the pull request:
`v91.0.0-backport.0`⏩`v91.3.1`

⚠️ The largest set of changes in this PR that touch source code (as opposed to test code) are related to several **EuiDataGrid** redesigns, particularly around the toolbar, column cell headers, and cell actions. We **strongly** recommend QAing your EuiDataGrid usages, **especially** if you have custom CSS styling on data grid cells.

| Changes | Screencap |
|--------|--------|
| Cell actions and popover | <img src="https://github.com/elastic/kibana/assets/549407/6462d983-307f-4a3c-84b1-36d9b276c9a0" width="240" alt=""> |
| Column headers | <img src="https://github.com/elastic/kibana/assets/549407/3fd64a15-829a-48f3-9dba-9dae3c73e6b2" alt="" width="360"> |
| Toolbar | <img src="https://github.com/elastic/kibana/assets/549407/f876f6d7-635d-497a-b1e7-9daf4e6fd3e3" alt="" width="240"> | 

---

## [`v91.3.1`](https://github.com/elastic/eui/releases/v91.3.1)

**Bug fixes**

- Moved `EuiDataGrid`'s header cells' `dataGridHeaderCellActionButton` test subject attribute from to the clickable button, for easier E2E testing ([#7427](https://github.com/elastic/eui/pull/7427))
- Fixed `EuiBasicTable`/`EuiInMemoryTable` actions to correctly show as disabled when rows are being selected ([#7428](https://github.com/elastic/eui/pull/7428))

## [`v91.3.0`](https://github.com/elastic/eui/releases/v91.3.0)

- Added `esqlVis`, `pipeBreaks`, and `pipeNoBreaks` icon glyphs. ([#7399](https://github.com/elastic/eui/pull/7399))
- Updated `EuiDataGridSchemaDetector`'s comparator arguments to include entry indexes ([#7406](https://github.com/elastic/eui/pull/7406))

## [`v91.2.0`](https://github.com/elastic/eui/releases/v91.2.0)

- Added `endpoint` glyph to `EuiIcon` ([#7383](https://github.com/elastic/eui/pull/7383))

**Bug fixes**

- Fixed a bug with `EuiSelectable`s with custom `truncationProps`, where scrollbar widths were not being accounted for ([#7392](https://github.com/elastic/eui/pull/7392))

## [`v91.1.0`](https://github.com/elastic/eui/releases/tag/v91.1.0)

- Updated `EuiDataGrid` cell actions to display above cells instead of within them, to avoid content clipping issues ([#7343](https://github.com/elastic/eui/pull/7343))
- Updated `EuiDataGrid` cell expansion popovers to sit on top of cells instead of below/next to them ([#7343](https://github.com/elastic/eui/pull/7343))
- Updated `EuiListGroupItem` to render an external icon and screen reader affordance for links with `target` set to to `_blank` ([#7352](https://github.com/elastic/eui/pull/7352))
- Updated `EuiListGroupItem` with a new `external` prop, which allows enabling or disabling the new external link icon ([#7352](https://github.com/elastic/eui/pull/7352))
- Updated `EuiText` to no longer set any opinionated styles on child `<img>` tags - use `EuiImage` for image display within text instead ([#7360](https://github.com/elastic/eui/pull/7360))
- Improved `EuiBasicTable`/`EuiInMemoryTable`s mobile UI for custom actions ([#7361](https://github.com/elastic/eui/pull/7361))
- Added a new `EuiDataGridToolbarControl` subcomponent, which is useful for rendering your own custom `EuiDataGrid` toolbar buttons while matching the look of the default controls ([#7369](https://github.com/elastic/eui/pull/7369))
- Updated `EuiDataGrid`'s toolbar controls to show active/current counts in badges, and updated the Columns button icon ([#7369](https://github.com/elastic/eui/pull/7369))
- Updated `EuiButtonEmpty` to allow passing `false` to `textProps`, which allows rendering custom button content without an extra text wrapper ([#7369](https://github.com/elastic/eui/pull/7369))
- Updated `EuiDataGrid` column header cells to show the sort arrow after the heading text, instead of before ([#7371](https://github.com/elastic/eui/pull/7371))
- Updated `EuiDataGrid`'s column header actions icon from a chevron to `boxesVertical` ([#7371](https://github.com/elastic/eui/pull/7371))
- Updated the actions column in `EuiBasicTable` and `EuiInMemoryTable`s. Alongside `name`, the `description`, `href`, and `data-test-subj` properties now also accept an optional callback that the current `item` will be passed to ([#7373](https://github.com/elastic/eui/pull/7373))
- Updated `EuiContextMenuItem` with a new `toolTipProps` prop ([#7373](https://github.com/elastic/eui/pull/7373))
- `EuiSelectable` now allows configurable text truncation via `listProps.truncationProps` ([#7388](https://github.com/elastic/eui/pull/7388))
- `EuiTextTruncate` now supports a new `calculationDelayMs` prop for working around font loading or layout shifting scenarios ([#7388](https://github.com/elastic/eui/pull/7388))

**Bug fixes**

- Fixed incorrect `EuiPopover` positioning calculations when `hasArrow` was set to false ([#7343](https://github.com/elastic/eui/pull/7343))
- Fixed `EuiSuperSelect` to render options with falsy values (false, 0, and ''), but not nullish values (undefined or null) ([#7362](https://github.com/elastic/eui/pull/7362))
- Fixed `EuiSuperSelect`'s typing to allow non-string values (e.g., booleans or numbers) ([#7362](https://github.com/elastic/eui/pull/7362))
- Fixed `EuiDataGrid`'s numeric and currency column heading cells to be correctly right-aligned ([#7371](https://github.com/elastic/eui/pull/7371))
- Fixed `EuiBasicTable` and `EuiInMemoryTable` actions not showing tooltip descriptions when rendered in the all actions popover menu ([#7373](https://github.com/elastic/eui/pull/7373))
- Fixed missing underlines on `EuiContextMenu` link hover ([#7373](https://github.com/elastic/eui/pull/7373))
- Fixed visual text truncation of `EuiBreadcrumb`s with `popoverContent` ([#7375](https://github.com/elastic/eui/pull/7375))
- Fixed `EuiFormRow`s with `hasEmptyLabelSpace` being very slightly off in vertical alignment ([#7380](https://github.com/elastic/eui/pull/7380))

**Deprecations**

- Deprecated `EuiContextMenuItem`'s `toolTipTitle` prop. Use `toolTipProps.title` instead ([#7373](https://github.com/elastic/eui/pull/7373))
- Deprecated `EuiContextMenuItem`'s `toolTipPosition` prop. Use `toolTipProps.position` instead ([#7373](https://github.com/elastic/eui/pull/7373))

**Accessibility**

- Fixed custom `EuiBasicTable`/`EuiInMemoryTable` rendering nested interactive custom actions ([#7361](https://github.com/elastic/eui/pull/7361))
- Fixed `EuiBasicTable` and `EuiInMemoryTable` actions not correctly reading out action descriptions to screen readers ([#7373](https://github.com/elastic/eui/pull/7373))
- Fixed `EuiBasicTable` and `EuiInMemoryTable` primary actions not visibly appearing on keyboard focus ([#7373](https://github.com/elastic/eui/pull/7373))

cee-chen committed:
i18n updates

cee-chen committed:
Upgrade EUI to latest v91.3.1

cee-chen commented:
/ci

cee-chen committed:
[EuiButton] Fix new `textProps` typing

- `textProps` can now be false, so we need the specific falsy check instead of using optional chaining

cee-chen committed:
[EuiLink] External icon className snapshot updates

cee-chen committed:
[EuiText] Emotion className hash updates

- hash changed due to removed child `img` styles

cee-chen committed:
[EuiText] Remove serverless search `<p>` workaround

- now that we're no longer opinionatedly adding margins to `<img>` tags within EuiText

- see https://github.com/elastic/eui/issues/7356#issuecomment-1810843908

cee-chen committed:
[EuiBasicTable] Test changes to account for table action updates

cee-chen committed:
[EuiDataGrid] Test changes to account for toolbar UI updates

cee-chen commented:
/ci

cee-chen committed:
[EuiSuperSelect] Fix type issue

- the component now allows setting non-nullish falsy values (e.g. false, 0, etc) as option values, so specifying `string` is needed here

cee-chen committed:
[EuiLink] Test fix for external icon logic

- icon is still there, but the text is not if `target="_blank"` is not set

cee-chen committed:
[EuiDataGrid] Test fixes/updates for new column count copy/UI

cee-chen committed:
[EuiDataGrid] Fix test to account for new cell actions location

cee-chen committed:
[EuiDataGrid][ml] Fix custom headers with histogram charts to account for new redesign

+ replace theme token usage with new preferred Emotion-inferred theme, which involves adding an `emotion.d.ts` file (https://emotion.sh/docs/typescript#define-a-theme)

cee-chen commented:
/ci

cee-chen committed:
[EuiBasicTable] Update test assertion for disabled actions UX

- EuiBasicTable no longer disables the `...` icon if all actions are disabled, only if the row is selected
- this is effectively the same UX as the individual actions in the popover are still disabled, but users now know specifically what actions they don't have access to

cee-chen committed:
[ml] Update chart assertion percentage

- not totally sure why this changed slightly, but my best guess is the header redesign

cee-chen committed:
[EuiBasicTable] Fix/update test checking for disabled action when row is selected

- a previous test that was selecting rows was polluting subsequent tests

- to be clear, the previous UX was already that the actions would be disabled on selection - the EUI fix was to remove the disabled actions from the DOM, which the test was asserting on

cee-chen committed:
[EuiDataGrid] Remove CSS overrides for cell actions

- should no longer be necessary with new design

cee-chen committed:
[EuiDataGrid] Update CSS selector, question about usage

cee-chen committed:
[EuiDataGrid] Update row height CSS selectors

cee-chen committed:
[EuiDataGrid][security_solution] Update custom CSS vertical alignment to account for new cell DOM

cee-chen committed:
[EuiDataGrid][Cloud Security] Update custom CSS vertical alignment to account for new cell DOM

cee-chen commented:
/ci

cee-chen committed:
[EuiContextMenuItem] Remove deprecated prop usage in favor of new `toolTipProps`

cee-chen commented:
/ci

elasticmachine commented:
Pinging @elastic/eui-team (EUI)

cee-chen commented on x-pack/packages/ml/data_grid/components/data_grid.tsx:
```diff
@@ -35,18 +36,49 @@ import { isPopulatedObject } from '@kbn/ml-is-populated-object';
+    // When a histogram chart is present in the column header, absolutely position
+    // the sorting arrow and actions icon so that the chart can expand to the full width
+    // of the header
+    '.cellHeaderWithHistogramChart': {
```
@elastic/ml-ui - I had to update a good amount of custom CSS to account for the histogram chart in the column header and the header redesign (see https://github.com/elastic/eui/pull/7371 for more info, or the PR description for more screenshots). Here's what the final behavior looks like:

![ml](https://github.com/elastic/kibana/assets/549407/007032ce-77ac-47c9-858c-d66f192e66d1)

I'm not as familiar with your UI and its edge cases however, and would appreciate your team QAing it to make sure everything looks/works as expected.

cc @MichaelMarcialis and @andreadelrio just as an FYI on this, to let y'all know how the new design affects some custom implementations (or vice versa)

cee-chen commented on x-pack/packages/security-solution/data_table/components/data_table/index.tsx:
@elastic/security-threat-hunting-investigations The changes in this file (and in `alerts_table`) should maintain vertically centered content primarily for auto height datagrid cells. Custom line count row heights aren't vertically aligned, but I believe this was already the case in prod(?).

If you don't find this to be the case or if you think some alignment has regressed, please let me know.

cee-chen commented on x-pack/plugins/cloud_security_posture/public/components/cloud_security_data_table/use_styles.ts:
```diff
@@ -27,18 +27,20 @@ export const useStyles = () => {
-    & .euiDataGridRowCell__expandFlex {
+      // Vertically center content
+      display: flex;
       align-items: center;
```
@elastic/cloud-security-posture I was unable to figure out how to get vulnerabilities data set up locally to test/QA this datagrid, so I'm not sure if the changes here work as expected without regressions. I would really appreciate your team's help QAing this to ensure this looks correct to you.

cee-chen commented on x-pack/plugins/cloud_security_posture/public/pages/vulnerabilities/hooks/use_styles.ts:
```diff
@@ -45,17 +45,18 @@ export const useStyles = () => {
+        flex-grow: 1;
+      }
+    }
+    /* EUI QUESTION: Why is this being done via CSS instead of setting isExpandable: false in the columns API? */
+    & .euiDataGridRowCell__actions > .euiDataGridRowCell__expandCell {
+      display: none;
     }
```
@elastic/kibana-cloud-security-posture this CSS override concerns me a little because it could easily look broken with the new cell actions design. If the cells in this data grid should not be expandable, I would like to switch to setting that via `isExpandable` API and removing this `display: none`.

cee-chen commented on x-pack/plugins/serverless_search/public/application/components/connectors_callout.tsx:
@elastic/enterprise-search-frontend @TattdCodeMonkey - FYI, I went ahead and removed the `<p>` workaround that you added as part of a discussion we had earlier in https://github.com/elastic/eui/issues/7356#issuecomment-1810843908. Would super appreciate you pulling this down locally to QA and confirming that it's still vertically aligned with the icon as expected!

sphilipse replied:
LGTM!

cee-chen replied:
Thanks a million, Sander!

jughosta committed:
[Discover] Adjust control cell styles

jughosta committed:
[Discover] Adjust toolbar controls

jughosta committed:
[Discover] Adjust timestamp column header

jughosta reviewed (commented):
Hi @cee-chen!

For Discover grid we would need some adjustments. I created a separate PR with these changes: https://github.com/cee-chen/kibana/pull/25

ElenaStoeva reviewed (approved):
ILM and License management changes (which seem to be jest snapshot updates only) LGTM

nreese reviewed (approved):
kibana-presenation changes LGTM
code review only

kc13greiner reviewed (approved):
LGTM!

cee-chen committed:
[Discover] Snapshot updates

crespocarlos reviewed (approved):
LGTM

jughosta reviewed (approved):
Thanks! LGTM 👍

dej611 reviewed (approved):
Visualizations changes LGTM 👍

walterra reviewed (approved):
ML/Transform changes LGTM, did a local test of the updated mini histograms in data grid, thanks for looking into updating the CSS!

banderror reviewed (approved):
`x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/use_bulk_actions.tsx` LGTM, tested locally 👍

PhilippeOberti commented:
@cee-chen I noticed that now that we're showing the cell actions outside of the cell, it kinda conflicts with the tooltip. While it might not be a big deal (once the mouse moves out of the content, the tooltip disappears), I want to highlight this behavior in case we want to rethink/improve it

| Before  | After |
| ------------- | ------------- |
| https://github.com/elastic/kibana/assets/17276605/62d1021d-3a32-4e2f-9c46-f69852f896f2 | https://github.com/elastic/kibana/assets/17276605/6ca80382-2e57-45c5-91eb-1ceec0e2ef30  |

cee-chen committed:
Merge branch 'main' into eui-v91.3.0

cee-chen commented:
@PhilippeOberti Good catch! @MichaelMarcialis, any design thoughts/suggestions here? My gut reaction would be to strongly suggest moving the tooltip position for your content to `down`/below the cell, instead of `top`.

cee-chen committed:
Merge branch 'main' into eui-v91.3.0

PhilippeOberti commented:
> @PhilippeOberti Good catch! @MichaelMarcialis, any design thoughts/suggestions here? My gut reaction would be to strongly suggest moving the tooltip position for your content to use `position="bottom"` & align the tooltip below the cell, instead of above/`top`.

While it would solve some of the situations, when the table is rendered towards the bottom of your browser window (which for the alerts page is almost always) having in code display of the tooltip on the bottom of he cell would not work, as it would be defaulted back to the top position automatically because there isn't any space below...

cauemarcondes reviewed (approved):
I tested it on Profiling pages and it works fine.

jloleysens reviewed (approved):
Core related changes LGTM!

cee-chen committed:
Merge branch 'main' into eui-v91.3.0

cee-chen commented:
@elastic/kibana-design @elastic/security-defend-workflows @elastic/security-threat-hunting-investigations @elastic/response-ops @elastic/cloud-security-posture

Happy new year y'all! We'd really appreciate a PR review (and especially QA from the above security-related teams who use EuiDataGrid, and are significantly impacted by this upgrade). We will be shooting for an admin merge sometime Thursday or Friday EOW otherwise.

MichaelMarcialis commented:
> @PhilippeOberti Good catch! @MichaelMarcialis, any design thoughts/suggestions here? My gut reaction would be to strongly suggest moving the tooltip position for your content to use `position="bottom"` & align the tooltip below the cell, instead of above/`top`.

Thanks, @cee-chen. As discussed in today's EUI weekly meeting, I think the best path forward is to:

1. Utilize a tooltip position other than `top` for any tooltips within a data grid cell (as you also mentioned above).
2. Change the elements' z-index so that the tooltips will always appear behind the data grid cell actions. This is for the case that the tooltip is rendered in close proximity to the edge of the user's viewport, and forcing the tooltip to position itself to the top again. This will prevent the cell actions from being partially obscured by the tooltip. I don't believe the overlapping of the actions above the tooltip will cause any usability issues, as the tooltip's padding should prevent any covering of the text within.

gergoabraham reviewed (approved):
Defend Workflow related changes look good! 🚢 

a big thank you for the cypress test updates 🙇

cee-chen committed:
Merge remote-tracking branch 'upstream/main' into eui-v91.3.0

cee-chen committed:
[Security Solution] Improve visibility of new cell actions design
- when combined with tooltips

cee-chen commented:
@PhilippeOberti - so unfortunately the z-index approach we wanted to take doesn't work due to tooltips being portalled to the document body 😞 I settled for changing the position to `bottom` for most of the tooltips, and bumping up the offset of top-aligned tooltips so that actions are at least more visible:

![datagridtooltip](https://github.com/elastic/kibana/assets/549407/43a4044d-75e2-4835-b87b-6e7bcb11d756)

That's the best I can do in this PR for now - if your team still wants a better longer-term solution, please let us/@MichaelMarcialis know so we can brainstorm something!

cee-chen committed:
[Cloud Security Posture] EuiDataGrid tweaks

- restore left alignment of header cells

- fix vertical center alignment of cell contents

- add more left spacing to custom toolbar data

- give timestamp tooltips the same workaround for new cell actions design that Security timeline received

cee-chen committed:
Merge branch 'main' into eui-v91.3.0

opauloh commented on x-pack/plugins/cloud_security_posture/public/pages/vulnerabilities/hooks/use_styles.ts:
```diff
@@ -45,21 +45,24 @@ export const useStyles = () => {
+      }
     }
-    & .euiDataGridRowCell__actions > [data-test-subj='euiDataGridCellExpandButton'] {
+    /* EUI QUESTION: Why is this being done via CSS instead of setting isExpandable: false in the columns API? */
+    & .euiDataGridRowCell__actions > .euiDataGridRowCell__expandCell {
       display: none;
     }
```
Thanks for pointing this out! I actually wasn't aware about this option, I'll update it to use `isExpanded` in a follow up PR

cee-chen replied:
Thanks!! 🙇  I was able to take a peek at this when I was fixing the UI locally. `display: none` isn't visually breaking in this case, but it would be a very minor accessibility issue for non-keyboard screen reader users (since keyboard users can't access other cell actions except via the expansion popover).

If it's not a blocker, we'd love for this CSS to simply be removed, but we'll leave it up to you all/another PR!

opauloh reviewed (approved):
cloud security changes LGTM 🚀🚀🚀

PhilippeOberti reviewed (approved):
thanks @cee-chen for fixing the weird tooltip issue. The solution you have now seems actually pretty good!

ymao1 reviewed (approved):
Response Ops changes LGTM
